spec/jira/client_spec.rb

```diff
@@ -1,223 +1,226 @@
-      to_return(:status => 200, :body => "", :headers => {})
+# We have three forms of authentication with two clases to represent the client for those different authentications.
+# Some behaviours are shared across all three types of authentication.  these are captured here.
+RSpec.shared_examples 'Client Common Tests' do
+  it { is_expected.to be_a JIRA::Client }
+
+  it 'freezes the options once initialised' do
+    expect(subject.options).to be_frozen
+  end
+
+  it 'prepends context path to rest base_path' do
+    options = [:rest_base_path]
+    defaults = JIRA::Client::DEFAULT_OPTIONS
+    options.each { |key| expect(subject.options[key]).to eq(defaults[:context_path] + defaults[key]) }
   end
 
-  let(:oauth_client) do
-    JIRA::Client.new({ :consumer_key => 'foo', :consumer_secret => 'bar' })
+  it 'merges headers' do
+    expect(subject.send(:merge_default_headers, {})).to eq({'Accept' => 'application/json'})
   end
 
-  let(:basic_client) do
-    JIRA::Client.new({ :username => 'foo', :password => 'bar', :auth_type => :basic })
+  describe 'http methods' do
+    it 'merges default headers' do
+      # stubbed response for generic client request method
+      expect(subject).to receive(:request).exactly(5).times.and_return(successful_response)
+
+      # response for merging headers for http methods with no body
+      expect(subject).to receive(:merge_default_headers).exactly(3).times.with({})
+
+      # response for merging headers for http methods with body
+      expect(subject).to receive(:merge_default_headers).exactly(2).times.with(content_type_header)
+
+      [:delete, :get, :head].each { |method| subject.send(method, '/path', {}) }
+      [:post, :put].each {|method| subject.send(method, '/path', '', content_type_header)}
+    end
+
+    it 'calls the generic request method' do
+      [:delete, :get, :head].each do |method|
+        expect(subject).to receive(:request).with(method, '/path', nil, headers).and_return(successful_response)
+        subject.send(method, '/path', {})
+      end
+
+      [:post, :put].each do |method|
+        expect(subject).to receive(:request).with(method, '/path', '', merged_headers)
+        subject.send(method, '/path', '', {})
+      end
+    end
   end
 
-  let(:cookie_client) do
-    JIRA::Client.new({ :username => 'foo', :password => 'bar', :auth_type => :cookie })
+  describe 'Resource Factories' do
+    it 'gets all projects' do
+      expect(JIRA::Resource::Project).to receive(:all).with(subject).and_return([])
+      expect(subject.Project.all).to eq([])
+    end
+
+    it 'finds a single project' do
+      find_result = double
+      expect(JIRA::Resource::Project).to receive(:find).with(subject, '123').and_return(find_result)
+      expect(subject.Project.find('123')).to eq(find_result)
+    end
   end
+end
 
-  let(:clients) { [oauth_client, basic_client, cookie_client] }
+RSpec.shared_examples 'HttpClient tests' do
```

Comment:
Good catch!  And very sorry that slipped through.  I thinking in trying to be clever (separating out a commit just with the spec restructure from the other changes) I managed to loose the invocation of the HttpClient shared examples for basic and cookie authentication.  I have resubmitted the PR with the two missing lines added.